### ISSUE TITLE
Expose `CUSTOM_AUTH` flow

### ISSUE THREAD

**Issue by kirillzh** [Liftric/cognito-idp#61]
I would like to use a [custom authentication flow](https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-authentication-flow.html#amazon-cognito-user-pools-custom-authentication-flow). Unfortunately current client API doesn't allow for that. There's `signIn` auth flow which is implemented like so:

```kotlin
override suspend fun signIn(
    username: String,
    password: String
): Result<SignInResponse> = request(
    Request.SignIn,
    SignIn(
        AuthFlow = Authentication.UserPasswordAuth.flow,
        ClientId = configuration.clientId,
        AuthParameters = SignIn.Parameters(username, password)
    )
)
```

Here `AuthFlow` is hardcoded to `USER_PASSWORD_AUTH`. If I'm not missing anything, adding a custom auth API should be as simple as:
1. Add `CognitoIdpClient#customAuth` API
2. Add new auth type for custom auth flow:
```kotlin
internal enum class Authentication(val flow: String) {
  RefreshTokenAuth("REFRESH_TOKEN_AUTH"),
  UserPasswordAuth("USER_PASSWORD_AUTH"),
  CustomAuth("CUSTOM_AUTH")
}
```
3. Use it in `customAuth` implementation:
```kotlin
override suspend fun customAuth(
  username: String,
  password: String,
): Result<SignInResponse> = request(
  Request.SignIn,
  SignIn(
    AuthFlow = CustomAuth.flow,
    ClientId = configuration.clientId,
    AuthParameters = SignIn.Parameters(username, password)
  )
)
```

**Comment by benjohnde** [comment]
Hi there, feel free to implement it and open a pull request :)

**Comment by kirillzh** [comment]
> Hi there, feel free to implement it and open a pull request :)

Submitted PR :)

**Comment by benjohnde** [comment]
@kirillzh should be released as `2.9.0`, release may take some hours to be published in a clear way

**Comment by benjohnde** [comment]
Thanks for the contribution once again :)
